File: support/verifygo/verifyProto.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"sync"

	sheeterProto "github.com/yinweli/Sheeter/support/verifygo/target/proto/codeGo"
)

func verifyProtoFrom(threads int) {
	loader := newProtoFileLoader()
	depot := sheeterProto.NewDepot(loader)

	assertProto(depot.FromData())
	verifyProto(depot, threads)

	fmt.Println("verify proto from: success")
}

func verifyProtoMerge(threads int) {
	loader := newProtoFileLoader()
	depot := sheeterProto.NewDepot(loader)

	assertProto(depot.MergeData())
	verifyProto(depot, threads)

	fmt.Println("verify proto merge: success")
}

func verifyProto(depot *sheeterProto.Depot, threads int) {
	waitGroup := &sync.WaitGroup{}
	waitGroup.Add(threads)

	for t := 0; t < threads; t++ {
		go func() {
			for i := int64(1); i <= 100; i++ {
				actual1, ok := depot.VerifyData1.Get(i)
				assertJson(ok)
				assertJson(actual1.GetKey() == i)
				assertJson(actual1.GetHide() == false)
				assertJson(actual1.GetEnable() == (i%2 == 1))
				assertJson(actual1.GetName() == fmt.Sprintf("名稱%d", i))
				assertJson(actual1.GetReward().GetDesc() == fmt.Sprintf("獎勵說明%d", i))
				assertJson(actual1.GetReward().GetGold() == i*2)
				assertJson(actual1.GetReward().GetDiamond() == i*3)
				assertJson(actual1.GetReward().GetCrystal() == i*4)
				assertJson(actual1.GetReward().GetFelIron() == i*5)
				assertJson(actual1.GetReward().GetAtium() == i*6)
				assertJson(len(actual1.GetReward().GetItem()) == 3)
				assertJson(actual1.GetReward().GetItem()[0].GetItemID() == 1000+i)
				assertJson(actual1.GetReward().GetItem()[0].GetType() == 0)
				assertJson(actual1.GetReward().GetItem()[0].GetCount() == i)
				assertJson(actual1.GetReward().GetItem()[1].GetItemID() == 10000+i)
				assertJson(actual1.GetReward().GetItem()[1].GetType() == 1)
				assertJson(actual1.GetReward().GetItem()[1].GetCount() == i)
				assertJson(actual1.GetReward().GetItem()[2].GetItemID() == 100000+i)
				assertJson(actual1.GetReward().GetItem()[2].GetType() == 2)
				assertJson(actual1.GetReward().GetItem()[2].GetCount() == i)

				actual2, ok := depot.VerifyData2.Get(i)
				assertJson(ok)
				assertJson(actual2.GetKey() == i)
				assertJson(actual2.GetHide() == false)
				assertJson(actual2.GetEnable() == (i%2 == 1))
				assertJson(actual2.GetName() == fmt.Sprintf("名稱%d", i))
				assertJson(actual2.GetReward().GetDesc() == fmt.Sprintf("獎勵說明%d", i))
				assertJson(actual2.GetReward().GetGold() == i*2)
				assertJson(actual2.GetReward().GetDiamond() == i*3)
				assertJson(actual2.GetReward().GetCrystal() == 0)
				assertJson(actual2.GetReward().GetFelIron() == 0)
				assertJson(actual2.GetReward().GetAtium() == 0)
				assertJson(len(actual2.GetReward().GetItem()) == 3)
				assertJson(actual2.GetReward().GetItem()[0].GetItemID() == 1000+i)
				assertJson(actual2.GetReward().GetItem()[0].GetType() == 0)
				assertJson(actual2.GetReward().GetItem()[0].GetCount() == i)
				assertJson(actual2.GetReward().GetItem()[1].GetItemID() == 10000+i)
				assertJson(actual2.GetReward().GetItem()[1].GetType() == 1)
				assertJson(actual2.GetReward().GetItem()[1].GetCount() == i)
				assertJson(actual2.GetReward().GetItem()[2].GetItemID() == 100000+i)
				assertJson(actual2.GetReward().GetItem()[2].GetType() == 2)
				assertJson(actual2.GetReward().GetItem()[2].GetCount() == i)
			} // for

			_, ok := depot.VerifyData1.Get(101)
			assertJson(ok == false)

			_, ok = depot.VerifyData2.Get(101)
			assertJson(ok == false)

			waitGroup.Done()
		}()
	} // for

	waitGroup.Wait()
}

type protoFileLoader struct {
	path string
}

func newProtoFileLoader() *protoFileLoader {
	_, root, _, ok := runtime.Caller(0)

	if ok == false {
		panic(fmt.Errorf("root path failed"))
	} // if

	return &protoFileLoader{
		path: filepath.Join(filepath.Dir(root), "target", "proto", "data"),
	}
}

func (this *protoFileLoader) Error(name string, err error) {
	panic(fmt.Errorf("%s: proto file load failed: %w", name, err))
}

func (this *protoFileLoader) Load(name, ext, fullname string) []byte {
	path := filepath.Join(this.path, fullname)
	data, err := os.ReadFile(path)

	if err != nil {
		panic(fmt.Errorf("%s: proto file load failed: %w", name, err))
	}

	return data
}

func assertProto(condition bool) {
	if condition == false {
		panic(fmt.Errorf("verify proto: verify failed"))
	} // if
}

File: support/verifyunity/Assets/VerifyProto.cs
using SheeterProto;
using System;
using System.Collections.Generic;
using System.IO;
using System.Threading;
using UnityEngine;

namespace verifycs
{
    public static class VerifyProto
    {
        public static void verifyProtoFrom(int threads)
        {
            var loader = new ProtoFileLoader();
            var depot = new Depot() { Loader = loader };

            assert(depot.FromData());
            verify(depot, threads);

            Debug.Log("verify proto from: success");
        }

        public static void verifyProtoMerge(int threads)
        {
            var loader = new ProtoFileLoader();
            var depot = new Depot() { Loader = loader };

            assert(depot.MergeData());
            verify(depot, threads);

            Debug.Log("verify proto merge: success");
        }

        private static void verify(Depot depot, int threads)
        {
            var threadList = new List<Thread>();

            for (var t = 0; t < threads; t++)
            {
                threadList.Add(
                    new Thread(() =>
                    {
                        VerifyData1 actual1;
                        VerifyData2 actual2;

                        for (var i = 1; i <= 100; i++)
                        {
                            assert(depot.VerifyData1.TryGetValue(i, out actual1));
                            assert(actual1 != null);
                            assert(actual1.Key == i);
                            assert(actual1.Hide == false);
                            assert(actual1.Enable == (i % 2 == 1));
                            assert(actual1.Name == "名稱" + i);
                            assert(actual1.Reward.Desc == "獎勵說明" + i);
                            assert(actual1.Reward.Gold == i * 2);
                            assert(actual1.Reward.Diamond == i * 3);
                            assert(actual1.Reward.Crystal == i * 4);
                            assert(actual1.Reward.FelIron == i * 5);
                            assert(actual1.Reward.Atium == i * 6);
                            assert(actual1.Reward.Item.Count == 3);
                            assert(actual1.Reward.Item[0].ItemID == 1000 + i);
                            assert(actual1.Reward.Item[0].Type == 0);
                            assert(actual1.Reward.Item[0].Count == i);
                            assert(actual1.Reward.Item[1].ItemID == 10000 + i);
                            assert(actual1.Reward.Item[1].Type == 1);
                            assert(actual1.Reward.Item[1].Count == i);
                            assert(actual1.Reward.Item[2].ItemID == 100000 + i);
                            assert(actual1.Reward.Item[2].Type == 2);
                            assert(actual1.Reward.Item[2].Count == i);

                            assert(depot.VerifyData2.TryGetValue(i.ToString(), out actual2));
                            assert(actual2 != null);
                            assert(actual2.Key == i.ToString());
                            assert(actual2.Hide == false);
                            assert(actual2.Enable == (i % 2 == 1));
                            assert(actual2.Name == "名稱" + i);
                            assert(actual2.Reward.Desc == "獎勵說明" + i);
                            assert(actual2.Reward.Gold == i * 2);
                            assert(actual2.Reward.Diamond == i * 3);
                            assert(actual2.Reward.Crystal == 0);
                            assert(actual2.Reward.FelIron == 0);
                            assert(actual2.Reward.Atium == 0);
                            assert(actual2.Reward.Item.Count == 3);
                            assert(actual2.Reward.Item[0].ItemID == 1000 + i);
                            assert(actual2.Reward.Item[0].Type == 0);
                            assert(actual2.Reward.Item[0].Count == i);
                            assert(actual2.Reward.Item[1].ItemID == 10000 + i);
                            assert(actual2.Reward.Item[1].Type == 1);
                            assert(actual2.Reward.Item[1].Count == i);
                            assert(actual2.Reward.Item[2].ItemID == 100000 + i);
                            assert(actual2.Reward.Item[2].Type == 2);
                            assert(actual2.Reward.Item[2].Count == i);
                        } // for

                        assert(depot.VerifyData1.TryGetValue(101, out actual1) == false);
                        assert(actual1 == null);

                        assert(depot.VerifyData2.TryGetValue(101.ToString(), out actual2) == false);
                        assert(actual2 == null);
                    })
                );
            } // for

            foreach (var itor in threadList)
                itor.Start();

            foreach (var itor in threadList)
                itor.Join();
        }

        private static void assert(bool condition)
        {
            if (condition == false)
                throw new Exception("verify proto: verify failed");
        }
    }

    class ProtoFileLoader : Loader
    {
        public void Error(string name, string message)
        {
            Debug.Log(name + ": proto file load failed: " + message);
        }

        public byte[] Load(FileName filename)
        {
            return File.ReadAllBytes(Path.Combine("Assets", "target", "proto", "data", filename.File));
        }
    }
}
